src/types.ts

```diff
@@ -529,36 +539,43 @@ const emailRegex =
 // }
 
 // Adapted from https://stackoverflow.com/a/3143231
-const datetimeRegex = (args: { precision: number | null; offset: boolean }) => {
-  if (args.precision) {
-    if (args.offset) {
-      return new RegExp(
-        `^\\d{4}-\\d{2}-\\d{2}T\\d{2}:\\d{2}:\\d{2}\\.\\d{${args.precision}}(([+-]\\d{2}:?\\d{2})|Z)$`
-      );
-    } else {
-      return new RegExp(
-        `^\\d{4}-\\d{2}-\\d{2}T\\d{2}:\\d{2}:\\d{2}\\.\\d{${args.precision}}Z$`
-      );
-    }
-  } else if (args.precision === 0) {
-    if (args.offset) {
-      return new RegExp(
-        `^\\d{4}-\\d{2}-\\d{2}T\\d{2}:\\d{2}:\\d{2}(([+-]\\d{2}:?\\d{2})|Z)$`
-      );
-    } else {
-      return new RegExp(`^\\d{4}-\\d{2}-\\d{2}T\\d{2}:\\d{2}:\\d{2}Z$`);
-    }
-  } else {
-    if (args.offset) {
-      return new RegExp(
-        `^\\d{4}-\\d{2}-\\d{2}T\\d{2}:\\d{2}:\\d{2}(\\.\\d+)?(([+-]\\d{2}:?\\d{2})|Z)$`
-      );
-    } else {
-      return new RegExp(
-        `^\\d{4}-\\d{2}-\\d{2}T\\d{2}:\\d{2}:\\d{2}(\\.\\d+)?Z$`
-      );
+const datetimeRegex = (args: {
+  precision?: number | null;
+  offset?: boolean;
+  withDate?: boolean;
+  withTime?: boolean;
+}) => {
+  let regex = "^";
+
+  if (args.withDate) {
+    regex = `${regex}\\d{4}-\\d{2}-\\d{2}`;
+
+    if (args.withTime) {
+      regex = `${regex}T`;
     }
   }
+
+  if (!args.withTime) {
+    return new RegExp(`${regex}$`);
+  }
+
+  regex = `${regex}\\d{2}:\\d{2}:\\d{2}`;
+
+  if (args.precision) {
+    regex = `${regex}\\.\\d{${args.precision}}`;
+  } else if (args.precision == null) {
```

Comment:
is the `==` intended? if so, could this be just `} else {`? I think it depends if we want to implicitly allow arbitrary precision

Reply:
> is the == intended? if so, could this be just } else {?

This is intentional to support both `precision = 0` (no dot, no anything) and `precision = null` OR `precision = undefined` (the don't care arbitrary default mode). It can't be a simple `else` since it'll catch the 0 case, add a `.` and allow arbitrary precision.

I'll add a comment and a test case for the 0 case.

> I think it depends if we want to implicitly allow arbitrary precision

Currently `.datetime()` defaults to arbitrary precision so had to align.